Feature form improvements [QgsQuick]

sklencar opened the pull request:
Changes are related to following widgets used in the FeatureForm:

Checkbox:
* fixed old import of QgsQuick
* changed indicator of a checkbox to be able to apply styles

DateTime:
* fixed wrong behaviour of the widget if a value is undefined
* replaced label with textfield with formatting options (ported from QField)

ExternalResources:
* fixed pixalation of icons
* added new icon without alternative border to be able to put color overlay on it
* fixed broken layout of icons

ValueMap
* updated way how to read option values from field config/setup (ported from QField, related to changes of the field config since QGIS 3.x)

TextEdit:
* Unified styling of the widgets - added some padding, changed background, height

Fields style added to QgsFeatureFormStyle
Added "fields" style object to customize some properties (e.g color, height, radius) of editor widgets. Due to new style options also widgets have been updated.

## Description
Include a few sentences describing the overall goals for this PR (pull request). If applicable also add screenshots.

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [ ] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [ ] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] New unit tests have been added for core changes
- [ ] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit

Commit:
[QgsQuick] Feature form improvements
Changes are related to following widgets used in the FeatureForm:

Checkbox:
* fixed old import of QgsQuick
* changed indicator of a checkbox to be able to apply styles

DateTime:
* fixed wrong behaviour of the widget if a value is undefined
* replaced label with textfield with formatting options (ported from QField)

ExternalResources:
* fixed pixalation of icons
* added new icon without alternative border to be able to put color overlay on it
* fixed broken layout of icons

ValueMap
* updated way how to read option values from field config/setup (ported from QField, related to changes of the field config since QGIS 3.x)

TextEdit:
* Unified styling of the widgets - added some padding, changed background, height

Fields style added to QgsFeatureFormStyle
Added "fields" style object to customize some properties (e.g color, height, radius) of editor widgets. Due to new style options also widgets have been updated.

PeterPetrik commented on src/quickgui/plugin/editor/qgsquickdatetime.qml:
```diff
@@ -25,101 +25,171 @@ import QgsQuick 0.1 as QgsQuick
+                inputMethodHints: Qt.ImhDigitsOnly
+
+                // this is a bit difficult to auto generate input mask out of date/time format using regex
+                // mainly because number of caracters is variable (e.g. "d": the day as number without a leading zero)
```
typo characters

PeterPetrik commented on src/quickgui/plugin/editor/qgsquickexternalresource.qml:
```diff
@@ -39,36 +43,57 @@ Item {
     function getSource() {
       if (image.status === Image.Error)
-        return QgsQuick.Utils.getThemeIcon("ic_broken_image_black")
+        return ""
```
should it be 

if (image.currentValue ...)
else 
return ""

?

Also now you have separate image for broken/missing and separate one for working version? Why is that?

sklencar replied:
Reverted back to one image for icons/preview photo

PeterPetrik commented on src/quickgui/plugin/editor/qgsquickvaluemap.qml:
```diff
@@ -25,35 +25,63 @@ import QgsQuick 0.1 as QgsQuick
+        }
+        else
+        {
+          //it's a map (<=QGIS2.18)
```
Do we need to support 2.18?

sklencar replied:
Its ported from QField and I would keep it.

PeterPetrik commented on src/quickgui/plugin/qgsquickfeatureformstyling.qml:
```diff
@@ -44,4 +44,16 @@ QtObject {
+      property real height: 48 * QgsQuick.Utils.dp
+      property real cornerRadius: 0 * QgsQuick.Utils.dp
+      property int fontPixelSize: 48 * QgsQuick.Utils.dp
+
```
extra line

PeterPetrik commented on src/quickgui/plugin/qgsquickphotopanel.qml:
```diff
@@ -27,14 +27,17 @@ Drawer {
 
-  property color bgColor: "white"
-  property real bgOpacity: 0.8
+  property color bgColor: "black"
```
should this be in config too?

sklencar replied:
I would keep it for time-being. PhotoPanel could have its own group in qgsquickfeatureformstyling which can be added later. It could contain properties such as icons, coloring, layout of buttons.

PeterPetrik reviewed (commented):
maybe would be nice to get opinion of @m-kuhn ?

m-kuhn commented on src/quickgui/plugin/qgsquickphotopanel.qml:
```diff
@@ -87,28 +88,27 @@ Drawer {
+          focusMode: Camera.FocusMacro
+          focusPointMode: Camera.FocusPointCustom
+          customFocusPoint: Qt.point(0.2, 0.2) // Focus relative to top-left corner
+      }
```
Is this here on purpose or a leftover from an experiment?

sklencar replied:
Added a fix in a recent commit. Focus hasn't been working in the previous version.

m-kuhn commented:
Good stuff in here.
But quite hard to review as it's quite a lot of different things. Have they gone to some other (modular) review before at lutra?

PeterPetrik commented:
if I understand it correctly it is tested with latest (development) version of https://github.com/lutraconsulting/input ( @wonder-sk ?)

m-kuhn commented:
Evenmore, the bugfix label is probably only partially applicable given the above ;)

wonder-sk commented:
> Have they gone to some other (modular) review before at lutra?

unfortunately not :-/ But we discussed already with @sklencar that next time the individual fixes should go to separate well defined commits...

> Evenmore, the bugfix label is probably only partially applicable given the above ;)

There are no new features here really, so I would be happy to get this merged in 3.6 - but if others feel strongly against we can postpone it to 3.8...

m-kuhn commented:
> unfortunately not :-/ But we discussed already with @sklencar that next time the individual fixes should go to separate well defined commits...

Great :+1: 

> There are no new features here really, so I would be happy to get this merged in 3.6 - but if others feel strongly against we can postpone it to 3.8...

Not really, I'm in favor of merging quickly.
And in general I don't think we should treat the whole quick library with the same level of stability as the rest. There are very few dependencies on it and no official downloads will ship it, it's really just a dev library. I'd actually be in favor of tagging it as experimental to make this clear, including no API stability guarantee.

Commit:
Some changes of widgets after review:
PhotoPanel: added focus
TextEdit: Fixed references for edit widget and removed redundant empty lines and properties, removed textWrap property for single line edit
ExternalRersources: Revert back to use one image for icons and previewPhoto with better handling
DateTime: fixed typos
